Initialize mapping support point weights through fast diagonalization method

kronbichler opened the pull request:
The constructor of `MappingQ` at degrees larger than 6-8 in 3D takes ages. We do not need that data structure very often as we now have a better alternative in `TransfiniteInterpolationManifold` that bypasses this logic or always use a volume manifold.

An alternative would be to write a better implementation for computing the Laplace matrix. It is a Laplace matrix on the unit cube with inner degrees of freedom, so one could invert it exactly with the fast diagonalization method (inversion of 1D matrices) and some better assembly commands.

bangerth commented:
Need to reindent.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -1140,6 +1114,42 @@ MappingQGeneric<dim,spacedim>::get_degree() const
+        output[2] = compute_support_point_weights_on_hex(polynomial_degree);
+
+      const_cast<std::vector<Table<2,double> > &>(support_point_weights_perimeter_to_interior)
+        .swap(output);
```
Instead of using `swap`, I think the more modern way of doing this is to use `dst = std::move(src)`.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -1140,6 +1114,42 @@ MappingQGeneric<dim,spacedim>::get_degree() const
+        output[2] = compute_support_point_weights_on_hex(polynomial_degree);
+
+      const_cast<std::vector<Table<2,double> > &>(support_point_weights_perimeter_to_interior)
+      .swap(output);
```
Instead of using swap, I think the more modern way of doing this is to use dst = std::move(src).

kronbichler replied:
I agree on this one.

kronbichler commented:
@bangerth I see your point but I have not considered it for two reasons. Firstly, I was following what we already do with the initialization of the prolongation matrices in the `FE_Q` and `FE_DGQ` classes. Secondly, I only got annoyed by this when I saw something that did not scale when increasing the resources with mappings of degrees higher than 10 in 3D on a large supercomputer where MPI takes all resources and I do not have background tasks - I usually compile even without threads there. So it would not help me, but that alone should of course not guide the overall decision. We might change that at some point, but I would not introduce a new mechanism now unless you really insist in that. (The more productive approach, if I had a bit more time, would be to implement a faster matrix inversion algorithm.)

bangerth commented:
OK, I figured as much (that you don't work with threads, and that consequently it wouldn't help you) :-)

I do in the future want to go to more lazy initializations in places where we create objects early and use them late. But that's a separate concern.

OK from my perspective.

kronbichler commented:
I pushed a completely different solution to this branch, namely the properly performing algorithm for computing the Laplace vector. It uses tensor product evaluation kernels for the high-order basis functions and the fast diagonalization method for computing the inverse of the Laplace matrix.

Here are some timings of the `compute_laplace_vector` function:
```
2D, degree 2:  old method 2e-5s,      new method 2.4e-4s
2D, degree 12: old method 0.023s,     new method 0.0041s
3D, degree 2:  old method 2e-5s,      new method 1.8e-4s
3D, degree 5:  old method 0.015s,     new method 0.0094s
3D, degree 8:  old method 0.90s,      new method 0.13s
3D, degree 12: old method 57s,        new method 1.4s
3D, degree 20: old method MANY HOURS, new method 29s
```
The new method is so fast that one can even compute very crazy mapping degrees that would be nearly impossible for the mapping infrastructure to handle in other places. I think we do not need to use the lazy initialization given this speed. Opinions?

I ran all mapping-related tests and there are no failures. Since this is a widely used feature, in particular `mappings/mapping_q_mixed_manifolds_0{1,2}`, we should have sufficient test coverage without a new test. The new computation should be even a bit more accurate since it does not invert a matrix (it was `long double` before, but still). The inaccuracy in the generalized eigenvalue problem in the fast diagonalization method is handled by an iterative refinement step. We could probably skip that and also the `long double` precision, which should speed up things by a factor of 5-10, but it is cheap enough this way.

drwells commented:
This is a 10-50x speed improvement; very nice! I agree that there is no reason to use a mutex and lazy initialization here.

I am not sure that I understand this algorithm: you invert the Laplace matrix by diagonalizing each 1D matrix, right?

kronbichler commented:
The speedup of course depends a bit on the degree.

When it comes to the algorithm: I goes back to Lynch et al. from 1964 [1] and it states for a matrix constructed as `A \otimes M + M \otimes A` (where A is the 1D Laplacian and M the 1D mass matrix) one can state the inverse as `S \otimes S (\Lambda \otimes I + I \otimes \Lambda) S^T \otimes S^T`, where `S` is a matrix containing the generalized eigenvectors of `A s = lambda B s` and `\Lambda` a diagonal matrix of the generalized eigenvalues. I will add this to the code.

[1] http://dl.acm.org/citation.cfm?id=2716130

bangerth commented:
@kronbichler -- I think you are suggesting this in your last sentence, but just in case I misunderstand: Can you add some comments that describe (i) why the matrix we consider here as the particular form you utilize, and (ii) the specific decomposition you use to compute its inverse, along with the link to the paper?

It's definitely an obscure enough approach that it's worth describing what is happening here.

kronbichler commented:
I now put the implementation of the fast diagonalization method into a new class, the `TensorProductMatrix` class residing in the `lac` subdirectory, accessed through the `apply_inverse` method. I assume we will use this class in a wider context in the future, so we might as well start with that matrix now. The class has all the basic functionality, but I would like to extend it a bit (or encourage others to extend it) by the following functions:
- Different matrices in different coordinate directions by vector of matrices
- Support for vectorized array
- Non-symmetric matrix A
(Those will be addressed by separate PRs.)

kronbichler commented:
@bangerth I added an extensive presentation of the algorithm to the class description of `TensorProductMatrix`, so please have a look there.

kronbichler commented:
I did some short tests with the new code and the whole constructor of `MappingQ` now runs in a very reasonable time, below 1 second up to degree 14 and below 10 seconds up to degree 21 in 3D. (Note that I actually skipped `long double` in favor of `double` because I do not see a point in computing more than 13 digits of auxiliary points. `MappingQGeneric` does not get its accuracy from the exact position of the support points - we used to have much worse positions before - but rather the polynomials that are defined through these points. A slight shift even of 1e-8 will not change any of the desirable properties - except of course the possible problem with convergence rates we discussed in length in #4404 and #3874, but that is not due some small shifts.

drwells commented:
I really like this new class (and I really do think one of the best things about limiting ourselves to quadrilaterals and hexahedra is that we can do tricks like this).

A small consistency complaint: We use `-1` as the 'check at runtime' template parameter for `project_matrix_free`: could we use the same value here instead of `0`?

kronbichler commented:
I will change it to `-1`. I just have to adjust the automatic deduction for the tensor product kernels that use `size-1` and in turn have a check for `-1`... ;-)

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+ * underlying LAPACK implementation supports only float and double numbers, so
+ * only these two types are currently supported.
+ *
+ * @tparam size Compile-time array lenghts. By default at -1, which means that
```
typo: should be 'lengths'

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+  TensorProductMatrix();
+
+  /**
+   * Initalizes the matrix to the given mass matrix $M$ and derivative matrix
```
typo: 'Initializes'

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+      mass_copy(i,j) = mass_matrix(i,j);
+  for (unsigned int i=0; i<derivative_matrix.m(); ++i)
+    for (unsigned int j=0; j<derivative_matrix.n(); ++j)
+      deriv_copy(i,j) = derivative_matrix(i,j);
```
Can't we just use `LAPACKFullMatrix<Number>::operator=(const FullMatrix &)` here?

kronbichler replied:
Yes, we can. I only tried with the constructor (that is not available) and then stopped looking. Thanks!

bangerth replied:
Or if we can't, maybe we should add this kind of operator to `LAPACKFullMatrix` :-)

bangerth replied:
Alternatively, even a copy constructor.

drwells replied:
I noticed that when I started using the LAPACK bindings two years ago and I keep forgetting to fix it. I should open an issue.

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
```
should be 'same as the other `apply_inverse` function'

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+ * @f}
+ * in 2D and
+ * @f{align*}{
+ * L &= A \otimes M \otimes + M \otimes A \otimes M + M \otimes M \otimes M
```
I think that you forgot to add a matrix to the first term: we have `A x M x`

drwells replied:
This should be `A x M x M + M x A x M + M x M x A`, right? The last term has three `M`s.

kronbichler replied:
Thanks for those two remarks. One gets blind with all those matrices... ;-)

drwells commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+ * and $\Lambda$ is the diagonal matrix representing the generalized
+ * eigenvalues $\lambda$. Note that the vectors $s$ are such that they
+ * simultaneously diagonalize $A$ and $M$, $S^{\mathrm S} A S = \Lambda$ and
+ * $S^{\mathrm S} B S = I$. This method of matrix inversion is called fast
```
These should be `\mathrm T`, right?

drwells commented:
I don't understand the matrix free kernels at all but I assume you know how to fix things. This looks good aside from the minor typos.

Another minor style comment: It seems odd that we have a default constructor and a reinit method. Would you mind adding another constructor that immediately forwards two matrix arguments to `reinit`?

bangerth commented on doc/news/changes/minor/20170711MartinKronbichler-b:
```diff
@@ -0,0 +1,6 @@
+Improved: The constructor of MappingQGeneric used a naive inversion of a cell
```
You know, I think this should actually be in the "major changes" section. This definitely ought to be a topic for the next release paper.

kronbichler replied:
Well, with two upvotes I move it there.

bangerth commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+ *
+ * Note that this class allows for two modes of usage. The first is a use case
+ * with run time constants for the matrix dimensions that is achieved by
+ * setting the optional template parameter for the size to -1. The second mode
```
In most of the rest of the library, we use `numbers::invalid_unsigned_int` to denote invalid values. Would that be possible here as well?

kronbichler replied:
The point is that we have the template specified as an `int`, not an `unsigned int`, so I think that is what we do elsewhere.

bangerth replied:
I know -- I might have implicitly suggested to also make the template argument `unsigned` ;-)

bangerth commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+ *
+ * @tparam Number Type of the underlying array elements. Note that the
+ * underlying LAPACK implementation supports only float and double numbers, so
+ * only these two types are currently supported.
```
Do you think we could check this via a `static_assert`? This way, you'd get an error right away, rather than only somewhere downstream.

kronbichler replied:
We might soon extend this a bit, so I would like to postpone it for the moment until we have decided on which types to support (we might also think about complex numbers that we get when `A` is non-symmetric). As you say, `LAPACKFullMatrix` will complain somewhere, which should be fine for the next month...

bangerth commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,371 @@
+  /**
+   * An array for temporary data.
+   */
+  mutable AlignedVector<Number> tmp_array;
```
Every mutable variable needs to have its corresponding mutex to be thread-safe.

kronbichler replied:
I see your point (and had actually thought about it). But I would like to avoid the mutex here because these are (very) small functions and thought the general use case to be a separate class for each thread. An alternative would be to have a thread local class variable for the temporary vector that we can query. How about that?

bangerth replied:
If every thread creates its own TensorProductMatrix object, then the mutex doesn't do much harm: there is never contention for the mutex and acquiring it is cheap. Are you sure that mutex acquisition cost would be relevant at all compared to the rest of what the function does?

As for thread-local variables: this only works if nowhere in your code path between first and last access of the variable are you calling anything that might involve task switches or anything that might invoke the TBB. That's sometimes difficult to establish because it might happen in functions you call from here -- e.g., something like `vector=0`.

kronbichler replied:
I agree, and have added the mutex to the code.

kronbichler commented:
Alright, I fixed the comments. Regarding the matrix free kernels: All the various `apply` methods do is to run through the degrees of freedom in different directions (given by the first template argument) and with transpose/non-transpose variants (second template argument) and adding/not adding (third argument). They are extensively tested by 250 matrix free tests and now we added six more plus the ones implicitly over the mapping... ;-)

bangerth commented on include/deal.II/lac/tensor_product_matrix.h:
```diff
@@ -0,0 +1,387 @@
+  Number *t = tmp_array.begin();
+
+  eval.template apply<0, true, false> (S, src, t);
+  if (dim == 1)
```
instead of this `if(dim==1)... else if (dim==2) ...`, it might be easier to read to just use a `switch` statement.

bangerth replied:
you might have to repeat the trailing `eval` statements but that might actually be easier to read anyway.

kronbichler replied:
I tried to minimize the number of times I write the `apply` function, but I agree that having nicely separated code is better. (At least until we implement a template recursion functionality for dimension-independent tensor products.)

drwells commented:
@bangerth we need the generalized symmetric eigenvalue solver from LAPACK for this to work: I don't think it is reasonable to reimplement this ourselves.

Do we have a non-LAPACK tester set up?

kronbichler commented:
I agree with @drwells that we do not want to re-implement the generalized eigenvalue solver. The other question regarding LAPACK is whether we could simply make LAPACK a required package. It is so ubiquitous nowadays that I cannot think of many users we would exclude that way.

kronbichler commented:
> In the finite element classes, there are a couple of places where we need to invert mass matrices, as far as I recall. (You probably know better than I where exactly that was.) Would these also be candidates for tensor product decomposition like you do here?

Yes sure, we could apply tensor product inverses in a number of other places, too. I will do whenever I find one and have the time to do so. For mass matrices, we might need another variant, though, namely a simpler version of `M \otimes M` rather than `M \otimes A + A \otimes M` approached here. Thinking about a simpler tensor product `M_1 \otimes M_2`, we might want to choose a different name for the current class, though: `TensorProductMatrix` is rather the former, whereas the derivative case could be `TensorProductDerivativeMatrix` or `TensorProductMatrixSum`. Opinions?

bangerth commented:
Yes, it's worthwhile getting the naming right now. Indeed, I think of `TensorProductMatrix` as something of the form `A \otimes B`. What you have is some kind of symmetric version of that. Maybe that ought to go into the name?

kronbichler commented:
We could argue that we create a "symmetric" sum of the terms akin the summation of the Laplacian as a sum of three derivative matrices. So `TensorProductMatrixSymmetricSum`? Let's wait a day or two and see if people have other ideas.

bangerth commented:
Yes, that name works for me.

kronbichler commented:
OK, I now changed the name to `TensorProductMatrixSymmetricSum`. I left the name of the file at `tensor_product_matrix.h` because we should at some point also add the usual tensor product matrix version `A \otimes B \otimes C` to this file.

This PR should be ready for the final review now.

bangerth commented:
Very nice stuff -- I'm excited that we're getting these non-trivial and non-obvious things into the library!
